main/src/library/ChannelImpl.flix

```diff
@@ -0,0 +1,379 @@
+//
+// Copyright 2018 Simon Meldahl Schmidt
+// Copyright 2018-2021 Jonathan Lindegaard Starup
+// Copyright 2021 Justin Fargnoli
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+namespace ChannelImpl {
+
+    ///
+    /// The tuple corresponds to (
+    ///   `id`,
+    ///   `channelLock` - the lock of the channel,
+    ///   `buffered` - indicates whether a handoff is needed or not.
+    ///       If you try to put an element in a channel that's full, you wait until there's space,
+    ///   `maxQueueSize` - the maximum number of elements allowed in `deque`. This is `max(1, channel size)`,
+    ///   `deque` - a deque of the channel elements,
+    ///   `waitingGetters` - a set of conditions that are waiting for an element.
+    ///       This set is cleared after each new element,
+    ///   `waitingSetters` - a condition that can notify threads of available space in `deque`.
+    /// )
+    /// The reason for the difference between `waitingGetters` and `waitingSetters` is the select expression.
+    /// A thread can only wait for one condition which is a problem when a select expression requires waiting for a
+    /// set of channels. Therefore the thread makes a condition that represents the select expression and hands that to
+    /// every channel involved.
+    ///
+    /// put is not supported in select so a single condition can be reused for all insertions.
+    ///
+    pub enum ChannelImpl {
+        case Mpmc(
+                Int64,
+                ReentrantLock,
+                Bool,
+                Int,
+                MutDeque[Boxed],
+                MutList[(ReentrantLock, Condition)],
+                Condition
+            )
+    }
+
+    ///
+    /// Creates a new channel if `bufferSize` is not negative.
+    ///
+    pub def new(bufferSize: Int): Result[ChannelImpl, String] & Impure =
```

Comment:
Who will use this? You will not be able to call it from Lowering because it returns `Result` which is incompatible with new channel.

Reply:
I am not sure how it works currently in Flix. When you create a channel do you have to specific its size as an integer or as an expression?

Reply:
The compiler wont use it yes. It was from the previous attempt to make it usable in code.

The size can be any expression.